pkg/cluster/majorversionupgrade.go

```diff
@@ -0,0 +1,92 @@
+package cluster
+
+import (
+	"fmt"
+
+	"github.com/zalando/postgres-operator/pkg/spec"
+	v1 "k8s.io/api/core/v1"
+)
+
+// VersionMap Map of version numbers
+var VersionMap = map[string]int{
+	"9.5": 90500,
+	"9.6": 90600,
+	"10":  100000,
+	"11":  110000,
+	"12":  120000,
+	"13":  130000,
+}
+
+// IsBiggerPostgresVersion Compare two Postgres version numbers
+func IsBiggerPostgresVersion(old string, new string) bool {
+	oldN, _ := VersionMap[old]
+	newN, _ := VersionMap[new]
+	return newN > oldN
+}
+
+// GetDesiredMajorVersionAsInt Convert string to comparable integer of PG version
+func (c *Cluster) GetDesiredMajorVersionAsInt() int {
+	return VersionMap[c.GetDesiredMajorVersion()]
+}
+
+// GetDesiredMajorVersion returns major version to use, incl. potential auto upgrade
+func (c *Cluster) GetDesiredMajorVersion() string {
+
+	if c.Config.OpConfig.MajorVersionUpgradeMode == "full" {
+		if IsBiggerPostgresVersion(c.Spec.PgVersion, c.Config.OpConfig.TargetMajorVersion) {
+			c.logger.Infof("overwriting configured major version %s to %s", c.Spec.PgVersion, c.Config.OpConfig.TargetMajorVersion)
+			return c.Config.OpConfig.TargetMajorVersion
+		}
+	}
+
+	return c.Spec.PgVersion
+}
+
+func (c *Cluster) majorVersionUpgrade() error {
+
+	if c.OpConfig.MajorVersionUpgradeMode == "off" {
+		return nil
+	}
+
+	desiredVersion := c.GetDesiredMajorVersionAsInt()
+
+	if c.currentMajorVersion >= desiredVersion {
+		c.logger.Infof("cluster version up to date. current: %d desired: %d", c.currentMajorVersion, desiredVersion)
+		return nil
+	}
+
+	pods, _ := c.listPods()
+	allRunning := true
+
+	var masterPod *v1.Pod
+
+	for _, pod := range pods {
+		ps, _ := c.patroni.GetMemberData(&pod)
+
+		if ps.State != "running" {
+			allRunning = false
+			c.logger.Infof("identified non running pod, potentially skipping major version upgrade")
```

Comment:
Judging from the [condition below](https://github.com/zalando/postgres-operator/pull/1372/files#diff-46d182d6fa2f77000bdf0906dc6d1ea032a074500f2eb07b3ce730384702e3afR78) the upgrade is always skipped if there is at least one non-running pod, which totally makes sense.  The word "potentially" is misleading in this case.